### ISSUE TITLE
Preprocessor pre-definitions and support for values?

### ISSUE THREAD

**Issue by vsajip** [javacc21/javacc21#75]
Given what you said in #74:

> Consider:
#if python
SCAN {some python code} =>
#elif java
SCAN {some java code} =>
#endif

Could this be arranged slightly differently? For example, 

```
#if CODEGEN==python
SCAN {some Python code} =>
#elif CODEGEN==csharp
SCAN {some C# code} =>
#else
SCAN {some Java code} =>
#endif
```
Then it would be possible to pre-define `CODEGEN` in JavaCC based on (for example) a command-line parameter used when invoking JavaCC to generate code in some specific language. I see that both the preprocessor and JavaCC parsers currently use a `Set<String>` for pre-defined symbols - presumably that could be a `Map<String, String>` instead, to support this way of doing things? Of course, one could use a different symbol rather than `CODEGEN`, perhaps with an underscore to indicate the symbol's special "reserved" status. I noticed that the preprocessor grammar already allows support for simple expressions, including `==` and `!=`.

Also, I presume one could also do e.g.

```
SCAN {
#if CODEGEN==python
  some Python code
#elif CODEGEN==csharp
  some C# code
#else
  some Java code
#endif
} =>
```
if that made sense in a particular case?

**Comment by revusky** [comment]
> Could this be arranged slightly differently? For example,...

Well, it could be arranged plenty differently. And yes, maybe another way would be better. Probably.... 

The thing about this is that the origin of this was me thinking about doing a C# grammar and the first thing one runs into, right at the *pre-lexical* level, is that, unlike Java, C# has a preprocessor. So I implemented the C# preprocessor basically, and then it immediately occurred to me that this would be very useful in JavaCC internal development, so I decided to leverage it for that.

So, my main thing on the C# preprocessor is not that it is necessarily, in an absolute sense, the perfect preprocessor. My thing is more that, since I intend to eventually get C# support going, and I need a C# grammar, and if JavaCC needs a preprocessor internally... since the project is severely undermanned, it makes sense to just use the very same preprocessor!

I guess the bottom line is that, unless maintaining a different preprocessor for C# and JavaCC internallly were to provide some really *major* benefit (hard to define precisely what "major" is, admittedly...) I would have a fairly strong tendency to just do it this way.

Of course, if you wanted to take ownership of the preprocessor piece, and maintain it yourself, then... but that said, I don't think it would be a very good prioritisation of your time, frankly.

**Comment by vsajip** [comment]
Given that
https://github.com/javacc21/javacc21/blob/8c9e27d0184aaf48bb0e7321e18bd1695de7a65a/examples/preprocessor/Preprocessor.javacc#L153-L154
together with
https://github.com/javacc21/javacc21/blob/8c9e27d0184aaf48bb0e7321e18bd1695de7a65a/examples/preprocessor/Preprocessor.javacc#L160
and
https://github.com/javacc21/javacc21/blob/8c9e27d0184aaf48bb0e7321e18bd1695de7a65a/examples/preprocessor/Preprocessor.javacc#L187-L198
it looks like the var=value option is already supported by the grammar. Doesn't it just need changing the return type of the `PP_Expression` and converting to the boolean sense when processing C#? It doesn't necessarily make sense to limit the JavaCC preprocessor to only booleans, just because C#'s does, as long as it can be made to work compatibly when used to process C# code. Which it could be, without too much work, I expect.

**Comment by revusky** [comment]
> It doesn't necessarily make sense to limit the JavaCC preprocessor to only booleans, just because C#'s does

Well, not necessarily, no. It seems that the designers of the C# language made the conscious decision to support the ability to turn on/off different sections of the template, and basically nothing else. They could have had a more sophisticated preprocessor and opted not to. So, the question becomes whether JavaCC really needs a more sophisticated preprocessor for some reason.

It seems to me that once you decide that the preprocessor symbols can be strings as well as just booleans, then one would naturally start thinking in terms of this as a generalized aliasing scheme. There is already a bit of aliasing going on, you might have noticed, like when you write:

       INJECT PARSER_CLASS {...}

it actually replaces PARSER_CLASS with the name of the generated parser class. Also CURRENT_NODE is an alias for whatever the variable name of the current node is. But, as things stand, there is a fixed set of these things that are supported internally, so it would seem natural that you could use the preprocessor to define your own aliases. But I don't know how generally useful that really is.

> Which it could be, without too much work, I expect.

Well, it's surely more work than what the legacy JavaCC project does in a typical year. But... that's just another way of saying that you're right, it isn't that much work!

Well, look, my position on this is that if this is important enough to you to do the work, then go ahead. Do it and I'll merge it. (And doing the work also means writing some minimum doc, like a dokuwiki page, and also a testcase or two.) 

BUT.... if it's only important enough to you to ask me to do the work, then.... :-)

So I leave it up to you. (Though I still tend to think that there are better ways for you to spend your time!)

**Comment by vsajip** [comment]
> Well, look, my position on this is that if this is important enough to you to do the work, then go ahead. Do it and I'll merge it [...] if it's only important enough to you to ask me to do the work, then.... :-)

That's all I was looking for, a willingness to consider merging. I'll take a crack at it :smile:

**Comment by vsajip** [comment]
> Well, not necessarily, no. It seems that the designers of the C# language made the conscious decision to support the ability to turn on/off different sections of the template, and basically nothing else. 

They didn't want the free-for-all that comes with the preprocessor in C & C++, and I can sympathize with that. That's not to say they didn't dial it too far back the other way. I can't think of any particular anti-pattern that's enabled by allowing the kind of change I mentioned - IMO it's things like token-pasting and the ability to substitute values into actual code that leads to some preprocessor-enabled monstrosities. As long as the functionality's just for enabling/disabling blocks of code based on conditions, that seems fine.

**Comment by vsajip** [comment]
Here's the first cut of changes to the preprocessor, let me know what you think: https://github.com/vsajip/javacc-preprocessor

**Comment by revusky** [comment]
(Drat, I was writing a reply and somehow lost my text. Oh well....)

> Here's the first cut of changes to the preprocessor, let me know what you think: https://github.com/vsajip/javacc-preprocessor

Well, yeah, it's pretty straightforward. One point though: I think it would be better (somewhat, but it's not that important, I grant) if, instead of `csharpMode` being a variable, if it was a preprocessor symbol itself. So, for example, here:  https://github.com/vsajip/javacc-preprocessor/blob/main/Preprocessor.javacc#L189-L197 you would have instead something like:

<CODE><PRE>
#if csharpMode
   [
         <PP_ASSIGN>
         <PP_SYMBOL> {value = lastConsumedToken.getImage();}
   ]
#endif
</PRE></CODE>

For one thing, that way, we're exercising the system more. 

Also, you may have noticed that there is a way of passing in preprocessor symbols on the command line, so you can write something like:

      java -jar javacc.jar -p foo,bar,baz MyGrammar.javacc

We now need it to be possible to write:

      java -jar javacc.jar -p foo=bar,baz=bat MyGrammar.javacc

The relevant point is here: https://github.com/javacc21/javacc21/blob/master/src/java/com/javacc/Main.java#L192-L202

(Yeah, I know that main() method is a bit messy.)

**Comment by vsajip** [comment]
> instead of csharpMode being a variable, if it was a preprocessor symbol itself

I made it a run-time rather than a compile-time option deliberately, so that the same exact code could be used to parse C# within JavaCC (when we get around to supporting that) and parse JavaCC source as well (making use of the `var=value` functionality).

> For one thing, that way, we're exercising the system more.

True, but we can easily cover that functionality with tests, which will be needed in any case.

> The relevant point is here

Sure, we can get to that when the time is right (i.e. when the underlying functionality is incorporated into JavaCC) - shouldn't take too long.

**Comment by revusky** [comment]
> I made it a run-time rather than a compile-time option deliberately, so that the same exact code could be used to parse C# within JavaCC

Well, I guess you have a point there. In any case, As you know doubt saw (since you accepted the invitation) I gave you access to the repository, so you can make the changes to the preprocessor directly. (Or anything else too, but maybe it would be better to be a bit conservative initially.)

As for using this internally, there is generally a need to *re-bootstrap* the jarfile, which basically amounts to dropping in the newer javacc.jar into the bin directory, which is something that I do fairly regularly. I'll frequently implement a new feature precisely because it would be useful in internal development and then re-bootstrap so that I can use it. You may have noticed the `full-test` target in the ant build.

         ant full-test

That actually runs the rebuilt/test twice, the first time with the current bin/javacc.jar and then the second time around replacing the jarfile that was just built in the bin directory and doing it all again.

**Comment by vsajip** [comment]
> so you can make the changes to the preprocessor directly. (Or anything else too, but maybe it would be better to be a bit conservative initially.)

I can commit changes to my fork and generate a pull request which you can review, we can iterate and you can merge if/when you're happy with it. That way there are (at least) two sets of eyeballs on my work, which there should be given that I am very new to the codebase. I might well miss some connection that's obvious to you.

**Comment by vsajip** [comment]
Fixed in #77.
